homeassistant/helpers/template.py

```diff
@@ -141,6 +145,34 @@ def async_render(self, variables: TemplateVarsType = None,
         except jinja2.TemplateError as err:
             raise TemplateError(err)
 
+    def render_with_collect(
+            self, variables: TemplateVarsType = None,
+            **kwargs) -> (Union[str, TemplateError],
+                          EntityFilter):
+        """Render given template and collect an entity filter."""
+        if variables is not None:
+            kwargs.update(variables)
+
+        return run_callback_threadsafe(
+            self.hass.loop, self.async_render_with_collect,
+            kwargs).result()
+
+    @callback
+    def async_render_with_collect(
+            self, variables: TemplateVarsType = None,
+            **kwargs) -> (Union[str, TemplateError],
+                          EntityFilter):
+        """Render the template and collect an entity filter."""
+        assert self.hass and _COLLECT_KEY not in self.hass.data
+        entity_collect = self.hass.data[_COLLECT_KEY] = EntityFilter()
+        try:
+            result = self.async_render(variables, **kwargs)
+            return (result, entity_collect)
+        except TemplateError as ex:
+            return (ex, entity_collect)
```

Comment:
This almost makes me wonder if we should store the entity filter class on the template class.

Reply:
Yeh I thought quite a bit about this one. That's not the right association. The filter is associated with the result of the template, not the template itself. If the template is rendered with different states, the filter is potentially different. The filter is also associated with the exception, if the input states change then it's possible that you won't have an exception or have a different one.

You could pass the filter as an out parameter, but that doesn't fit with the python way of doing things because you don't have pointer references and you'd have to construct the filter and pass it in, which is the wrong diffusion of responsibility.

The only other option is to return a `TemplateResult`, which then raises an exception when you use the `result` property. The tuple results though really amounts to the same thing.

Reply:
good point.

Reply:
Actually having a `TemplateResult` be returned might be a good idea anyhow, particularly if [arch#206](/home-assistant/architecture/issues/206) gets implemented. I might change this one to be like this regardless. Without arch 206 there was less motivation, but if we pass back the refresh limits as well then there's reason enough to do so.